zebra-state/src/service/read/difficulty.rs

```diff
@@ -18,33 +19,85 @@ use crate::{
             AdjustedDifficulty,
         },
         finalized_state::ZebraDb,
+        read::tree::history_tree,
         NonFinalizedState,
     },
-    GetBlockTemplateChainInfo,
+    BoxError, GetBlockTemplateChainInfo, HashOrHeight,
 };
 
+use crate::service::block_iter::Iter;
+
 /// Returns the [`GetBlockTemplateChainInfo`] for the current best chain.
 ///
 /// # Panics
 ///
-/// If we don't have enough blocks in the state.
-pub fn difficulty_and_time_info(
+/// - If we don't have enough blocks in the state.
+/// - If a consistency check fails `RETRIES` times.
+pub fn get_block_template_chain_info(
     non_finalized_state: &NonFinalizedState,
     db: &ZebraDb,
     tip: (Height, Hash),
     network: Network,
 ) -> GetBlockTemplateChainInfo {
-    let relevant_chain = any_ancestor_blocks(non_finalized_state, db, tip.1);
-    difficulty_and_time(relevant_chain, tip, network)
+    // Amount of retries for a consistency check.
+    const RETRIES: usize = 3;
+
+    let mut relevant_chain_and_history_tree_result =
+        relevant_chain_and_history_tree(non_finalized_state, db, tip.1);
+
+    for _ in 0..RETRIES {
+        if relevant_chain_and_history_tree_result.is_ok() {
+            break;
+        }
+
+        relevant_chain_and_history_tree_result =
+            relevant_chain_and_history_tree(non_finalized_state, db, tip.1);
+    }
+    // remove the unwrap somehow
+    let (relevant_chain, history_tree) =
+        relevant_chain_and_history_tree_result.expect("consistency check failed");
```

Comment:
I don't think we should panic here but i want to ask before trying something else.

Reply:
Getting 3 tip errors means that Zebra is still syncing lots of blocks, so the RPC should return an error.

This is what we do for the address RPCs:
https://github.com/ZcashFoundation/zebra/blob/343c5e68d4ba73c985e36333b0889757624bbb60/zebra-state/src/service/read/address/utxo.rs#L97-L102

Scroll down to see how we handle the result:

https://github.com/ZcashFoundation/zebra/blob/ce07f4d7af9eae3fd80c38ff3ecf5adb71b7b1b7/zebra-state/src/service.rs#L1501-L1515

Reply:
By what i understand of your comment, getting 3 tip errors means we are out of sync but we are checking this at https://github.com/ZcashFoundation/zebra/blob/main/zebra-rpc/src/methods/get_block_template_rpcs.rs#L324

It is not clear for me how to handle the error here. 

Here is what i think we can do if we have to return an error:

The response for the call is `ChainInfo(Option<GetBlockTemplateChainInfo>)`,so we can use `None` if there was an error and add a new arm to https://github.com/ZcashFoundation/zebra/blob/main/zebra-rpc/src/methods/get_block_template_rpcs.rs#L355.

Reply:
I think you might want to take this and push a commit as i will like to avoid the back and forth.

Reply:
Ok, I can make this change, but I'd like you to check if the chain history hash is correct.

Reply:
Fixed in commit 7cd21fc by returning a `Result` from the function and in the state response. (The RPC already handles state errors by turning them into RPC errors.)